core/trino-main/src/main/java/io/trino/operator/project/PageFieldsToInputParametersRewriter.java

```diff
@@ -42,80 +45,117 @@ private PageFieldsToInputParametersRewriter() {}
         {
             return fieldToParameter.computeIfAbsent(reference.getField(), field -> {
                 inputChannels.add(field);
+                if (unconditionallyEvaluated) {
+                    eagerlyLoadedChannels.add(reference.getField());
+                }
                 return nextParameter++;
             });
         }
 
         @Override
-        public RowExpression visitCall(CallExpression call, Void context)
+        public RowExpression visitCall(CallExpression call, Boolean unconditionallyEvaluated)
         {
             return new CallExpression(
                     call.getResolvedFunction(),
                     call.getArguments().stream()
-                            .map(expression -> expression.accept(this, context))
+                            .map(expression -> expression.accept(this, unconditionallyEvaluated))
                             .collect(toImmutableList()));
         }
 
         @Override
-        public RowExpression visitSpecialForm(SpecialForm specialForm, Void context)
+        public RowExpression visitSpecialForm(SpecialForm specialForm, Boolean unconditionallyEvaluated)
         {
+            if (unconditionallyEvaluated) {
+                switch (specialForm.getForm()) {
+                    case IF:
+                    case SWITCH:
+                    case BETWEEN:
+                    case AND:
+                    case OR:
+                    case COALESCE:
+                        return specialFormWithConditionallyEvaluatedArguments(specialForm);
+                    default:
+                        break;
+                }
+            }
             return new SpecialForm(
                     specialForm.getForm(),
                     specialForm.getType(),
                     specialForm.getArguments().stream()
-                            .map(expression -> expression.accept(this, context))
+                            .map(expression -> expression.accept(this, unconditionallyEvaluated))
                             .collect(toImmutableList()),
                     specialForm.getFunctionDependencies());
         }
 
         @Override
-        public RowExpression visitConstant(ConstantExpression literal, Void context)
+        public RowExpression visitConstant(ConstantExpression literal, Boolean unconditionallyEvaluated)
         {
             return literal;
         }
 
         @Override
-        public RowExpression visitLambda(LambdaDefinitionExpression lambda, Void context)
+        public RowExpression visitLambda(LambdaDefinitionExpression lambda, Boolean unconditionallyEvaluated)
         {
             return new LambdaDefinitionExpression(
                     lambda.getArgumentTypes(),
                     lambda.getArguments(),
-                    lambda.getBody().accept(this, context));
+                    lambda.getBody().accept(this, unconditionallyEvaluated));
```

Comment:
This really depends on the function being applied. It's possible that not all of the references in the lambda expression is used.

Reply:
I see, I've updated `visitCall` to now assume that all the inputs to lambda expression can be conditionally evaluated (i.e. keep existing behaviour of all blocks being lazily loaded).

Reply:
> This really depends on the function being applied. It's possible that not all of the references in the lambda expression is used.

@martint could you put an example here?

Should we just assume that lambda body is `unconditionallyEvaluated=false` and add a TODO?